storage/local/storage.go

```diff
@@ -336,67 +336,96 @@ func (s *memorySeriesStorage) NewPreloader() Preloader {
 	}
 }
 
-// FingerprintsForLabelMatchers implements Storage.
-func (s *memorySeriesStorage) FingerprintsForLabelMatchers(labelMatchers metric.LabelMatchers) clientmodel.Fingerprints {
+// FingerprintsForLabels implements Storage.
+func (s *memorySeriesStorage) FingerprintsForLabels(pairs ...metric.LabelPair) map[clientmodel.Fingerprint]struct{} {
 	var result map[clientmodel.Fingerprint]struct{}
-	for _, matcher := range labelMatchers {
+	for _, pair := range pairs {
 		intersection := map[clientmodel.Fingerprint]struct{}{}
-		switch matcher.Type {
-		case metric.Equal:
-			fps, err := s.persistence.fingerprintsForLabelPair(
-				metric.LabelPair{
-					Name:  matcher.Name,
-					Value: matcher.Value,
-				},
-			)
-			if err != nil {
-				log.Error("Error getting fingerprints for label pair: ", err)
-			}
-			if len(fps) == 0 {
-				return nil
-			}
-			for _, fp := range fps {
-				if _, ok := result[fp]; ok || result == nil {
-					intersection[fp] = struct{}{}
-				}
+		fps, err := s.persistence.fingerprintsForLabelPair(pair)
+		if err != nil {
+			log.Error("Error getting fingerprints for label pair: ", err)
+		}
+		if len(fps) == 0 {
+			return nil
+		}
+		for _, fp := range fps {
+			if _, ok := result[fp]; ok || result == nil {
+				intersection[fp] = struct{}{}
 			}
-		default:
-			values, err := s.persistence.labelValuesForLabelName(matcher.Name)
-			if err != nil {
-				log.Errorf("Error getting label values for label name %q: %v", matcher.Name, err)
+		}
+		if len(intersection) == 0 {
+			return nil
+		}
+		result = intersection
+	}
+	return result
+}
+
+// Matches returns a the metrics from storage that satisfy the given label matchers.
+func Matches(storage Storage, matchers ...*metric.LabelMatcher) map[clientmodel.Fingerprint]clientmodel.COWMetric {
+	var (
+		equals  []metric.LabelPair
+		filters []*metric.LabelMatcher
+	)
+	for _, lm := range matchers {
+		if lm.Type == metric.Equal && lm.Value != "" {
+			equals = append(equals, metric.LabelPair{
+				Name:  lm.Name,
+				Value: lm.Value,
+			})
+		} else {
+			filters = append(filters, lm)
+		}
+	}
+
+	var resfps map[clientmodel.Fingerprint]struct{}
+	// If we cannot make a pre-selection based on equality matchers, expanding the other matchers to labels
+	// and intersecting their fingerprints is still likely to be the best choice.
+	if len(equals) > 0 {
+		resfps = storage.FingerprintsForLabels(equals...)
```

Comment:
I understand why it can make sense but then we have to enforce the constraint that a set of label matchers must not have two matchers of type `equal` with the same label name.
It also requires us to create a bunch of new single-element label set maps for each `!=`, `=~`, and `!~` matcher which in turn feels non-idiomatic.

I feel like the storage is low level enough to step down to label pairs.